internal/campaigns/workspace.go

```diff
@@ -0,0 +1,94 @@
+package campaigns
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/sourcegraph/src-cli/internal/api"
+	"github.com/sourcegraph/src-cli/internal/campaigns/graphql"
+)
+
+// WorkspaceCreator implementations are used to create workspaces, which manage
+// per-changeset persistent storage when executing campaign steps and are
+// responsible for ultimately generating a diff.
+type WorkspaceCreator interface {
+	// Create should clone the given repository, and perform whatever other
+	// initial setup is required.
+	Create(context.Context, *graphql.Repository) (Workspace, error)
+
+	// DockerImages returns any Docker images required to use workspaces created
+	// by this creator.
+	DockerImages() []string
+}
+
+// Workspace implementations manage per-changeset storage when executing
+// campaign step.
+type Workspace interface {
```

Comment:
This is a really neat abstraction 👏 I've grown to be skeptical when I see interfaces (maybe even: abstractions in general? am I getting old?), but I think this one here falls in place just right.